Provide an update guide

rolfedh opened the pull request:
Fixes https://github.com/quarkusio/quarkus/issues/32585
Doc preview: https://quarkus-pr-main-32384-preview.surge.sh/version/main/guides/update-to-quarkus-3

michelle-purcell commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,89 @@
+////
+[id="update-projects-to-quarkus-3-howto"]
+= Update projects to Quarkus 3 automatically
+:extension-status: preview
```
The order and ID items in the header items differ from the [contributor guide template & example](https://quarkus.io/guides/doc-contribute-docs-howto#create-quarkus-content-in-asciidoc) for the new diataxis implementation.
 
Minor observation, not sure it makes a difference in upstream output, but it might for our downstream post-fix script/processes. The more consistent the easier for later.

rolfedh replied:
Updated. Thanks.

michelle-purcell commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,89 @@
+:summary: Update projects from Quarkus 2.13 and later to Quarkus 3.
+
+As a developer, you can update your projects from Quarkus 2.13 and later to Quarkus 3 by running one of the update commands.
+These commands use OpenRewrite recipes to automatically update most of your project's dependencies, source code, and documentation.
```
_These_ is ambiguous. Could we describe the commands and how you can run them earlier before the steps begin? 

Line 58 has some good information that could be placed next to this line.

`You can run the update command from either the Quarkus CLI, Maven, or JBang.
`

rolfedh replied:
Updated. Thanks.

michelle-purcell commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,89 @@
+
+== Procedure
+
+. If your project uses Hibernate ORM or Hibernate Reactive, see the link:https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0:-Hibernate-ORM-5-to-6-migration[Hibernate ORM 5 to 6 migration] quick reference.
```
Is the goal/action of step 1 to make sure you are using version 6 of Hibernate ORM? If so, should we be direct here about the required action. Also, wondering if this is more of a prereq?

rolfedh replied:
I'll ask @yrodiere for clarification about the best placement of this information.

michelle-purcell commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,89 @@
+[IMPORTANT]
+====
+After updating, evaluate whether your project requires additional follow-up steps.
+Verify that the application from your updated project works as expected before releasing it.
```
```suggestion
Verify that the application from your updated project works as expected before releasing it to production.
```

michelle-purcell reviewed (commented):
@rolfedh Nice work :+1:   See comments above. Thanks

rolfedh commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,89 @@
+https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
+////
+[id="update-projects-to-quarkus-3-howto"]
+= Update projects to Quarkus 3 automatically
```
I'd like to use the `{quarkus-version}` attribute but am concerned that for now, during this pre-release period, it would show the wrong version number and confuse reviewers. I've created [a follow-up task](https://issues.redhat.com/browse/QDOCS-295) to take care of it post-release.

rolfedh commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,89 @@
+. Use your version control system to create a working branch for your project or projects.
+. Update the project using the Quarkus CLI:
++
+[source,bash]
```
[The blog post](https://quarkus.io/blog/quarkus-3-0-0-cr2-released/#automated-migration) *recommends* the Quarkus CLI update command, but not the alternative commands. I'll get SME feedback on this question. If the commands ARE equivalent, I'll combine them as you've suggested.

gsmet replied:
The `quarkus update` and `mvn` command are equivalent. The JBang one should not be documented.

rolfedh replied:
Updated. Thanks.

rolfedh commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,98 @@
+
+[IMPORTANT]
+====
+If your project uses Hibernate ORM or Hibernate Reactive, see the link:https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0:-Hibernate-ORM-5-to-6-migration[Hibernate ORM 5 to 6 migration] quick reference.
```
@yrodiere 
What's the best placement for this note? 
Does the update command take care of any of the changes described in this quick reference?

yrodiere replied:
>  What's the best placement for this note?

I'd merge it into the IMPORTANT block above that starts with "After updating, evaluate whether your project requires additional follow-up steps"?

E.g. :

```
[IMPORTANT]
====
After updating, evaluate whether your project requires additional follow-up steps.

Verify that the application from your updated project works as expected before releasing it to production.

In particular, if your project uses Hibernate ORM or Hibernate Reactive, you will want to read through the link:https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0:-Hibernate-ORM-5-to-6-migration[Hibernate ORM 5 to 6 migration] quick reference.
====

>  Does the update command take care of any of the changes described in this quick reference?

Some, yes; most simply cannot be automated. 

The package changes (`javax.persistence` => `jakarta.persistence`) is definitely handled, and I think @gsmet was considering enabling the [Hibernate ORM 5.6 compatibility switch](https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0:-Hibernate-ORM-5-to-6-migration#best-effort-hibernate-orm-56-compatibility-switch), which should address most changes in the "Database schema" section, but I'm not sure he got to doing it.

Regardless... Surely Hibernate ORM is not the only library with changes that are not handled by `quarkus upgrade`. Shouldn't this problem be approached more globally for the whole migration guide? Like linking to the migration guide from here, and somehow documenting in the migration guide whether each section of the migration guide is handled by `quarkus upgrade` or not.

rolfedh replied:
Updated. Thanks.

rolfedh commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,98 @@
+
+[source,bash]
+----
+./mvnw io.quarkus.platform:quarkus-maven-plugin:3.0.0.Final:update -N -Dstream=3.0
```
Is `3.0.0.Final` the [correct tag](https://github.com/quarkusio/quarkus/tags) or should users specify a later tag?

gsmet replied:
Use `{quarkus-version}` instead. The idea is to use the latest one available.

rolfedh replied:
Updated. Thanks.

rolfedh commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,98 @@
+As a developer, you can update your projects from Quarkus 2.13 and later to Quarkus 3 by running an update command.
+The update command uses OpenRewrite recipes to automatically update most of your project's dependencies, source code, and documentation.
+
+// Give the link to the OpenRewrite recipe?
+// How does the user know what the update command does or does not cover?
+// In the migration guide, can we identify which items are/are not covered by the OpenRewrite recipe?
+// Do we want to invite feedback? For example: If it does not work for you, we could have missed something, or you are using extensions not yet supporting Quarkus 3.
+// https://github.com/quarkusio/quarkus/issues[Let us know] in either case so we can improve the migration script together.
```
Any thoughts on these questions?

gsmet replied:
> Give the link to the OpenRewrite recipe?

We could but it's hard to understand what it is doing at a glance as it's very long.

> In the migration guide, can we identify which items are/are not covered by the OpenRewrite recipe?

We hope to do that at some point but we are not there yet.

> Do we want to invite feedback?

Yes, please.

rolfedh replied:
Updated. Thanks.

rolfedh commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,98 @@
+. Verify that the project builds without errors and that the application passes all tests and works as expected before releasing it to production.
+
+[id="alternative-commands"]
+== Alternative commands
```
Do the following commands produce the same results as `quarkus update`? 
Should we keep them here, or would it be better to present them alongside the `quarkus update` command, above?

gsmet commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,98 @@
++
+[NOTE]
+====
+If you choose not to install Quarkus CLI version 3, you can update your project by using Maven or JBang commands in the xref:alternative-commands[Alternative commands] section.
```
The JBang script will go away, let's not document it.

rolfedh replied:
Updated. Thanks.

gsmet commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,98 @@
+. Review the output from the `quarkus update` command for potential instructions and, if needed, perform the indicated tasks.
+. Review all the changes using a diff tool.
+. Review link:https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0[the migration guide] for any items not covered by the upgrade command and perform additional steps, if needed.
+// IMPORTANT: As a developer deciding whether to use the update command, knowing which tasks the update command handles makes it more likely that I'll use this feature. Otherwise, I might avoid it.
```
I agree with that. I'm not sure I will have the time to do anything about it soon though.

rolfedh replied:
Updated. Thanks.

rolfedh replied:
I created https://github.com/quarkusio/quarkus/issues/32807 to explain and track this potential issue.

gsmet commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,98 @@
+//
+//I will update and uncomment this section if/when this capability becomes available.
+
+.JBang
+
+If you have JBang installed, update your project:
+[source,bash]
+----
+jbang --fresh upgrade-to-quarkus3@quarkusio
+----
+
+For Linux and macOS, if you don't have JBang installed, install JBang and update your project:
+[source,bash]
+----
+curl -Ls https://sh.jbang.dev | bash -s - --fresh upgrade-to-quarkus3@quarkusio
+----
+
+For Microsoft Windows, if you don't have JBang installed, install JBang and update the project:
+[source,bash]
+----
+iex "& { $(iwr https://ps.jbang.dev) } --fresh upgrade-to-quarkus3@quarkusio"
+----
```
As previously stated, let's drop this part.

rolfedh replied:
Updated. Thanks.

gsmet reviewed (changes requested):
Thanks! I suggested a few changes.

gsmet commented:
Also please make sure the title of the PR is descriptive enough. I fixed it.

maxandersen commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,55 @@
+
+include::{includes}/extension-status.adoc[]
+
+As a developer, you can update your projects from Quarkus 2.13 and later to Quarkus 3 by running an update command.
```
```suggestion
You can update your projects from Quarkus 2.x to Quarkus 3.x by running an update command.
```
a) do we need to be a developer to do this? can't we assume those reading this have developer skills?
 
b) why not just say Quarkus 2.x?

rolfedh replied:
Thanks. Updated. 
I had understood `quarkus update` worked only on 2.13 and later. 
@gsmet can you confirm that it works for Quarkus 2.7-based projects as well?

maxandersen commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,55 @@
+include::_attributes.adoc[]
+:categories: core
+:extension-status: "experimental"
+:summary: Update projects from Quarkus 2.13 and later to Quarkus 3.
```
The guide isn't just Quarkus 2.13 and later Quarkus 3 - the update tooling will also work for Quarkus 3.0-3.x's

maxandersen replied:
scratch that - now having  read it all i see this is specifically for quarkus 3 migration. not a generic explanation of the update toolling.

rolfedh replied:
Thank  you, Max.

rolfedh replied:
Thanks for your thoughts on this.

maxandersen commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,55 @@
+- The recipe might not cover an item in your project.
+- Your project might use an extension that does not support Quarkus 3 yet.
+
+In either case, https://github.com/quarkusio/quarkus/issues[let us know by filing an issue] so we can improve the update recipe.
```
```suggestion
In either case, https://github.com/quarkusio/quarkus/issues[let us know by filing an issue] so we can improve the update command.
```

its not just recipes that can be at fault; also the auxiliary code/commands used to implement it.

rolfedh replied:
Updated. Thanks.

maxandersen commented on docs/src/main/asciidoc/update-to-quarkus-3.adoc:
```diff
@@ -0,0 +1,55 @@
+
+As a developer, you can update your projects from Quarkus 2.13 and later to Quarkus 3 by running an update command.
+
+The update command uses OpenRewrite recipes to automatically update most of your project's dependencies, source code, and documentation. These recipes cover many but not all of the items described in the link:https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0[Migration Guide 3.0].
```
```suggestion
The update command primarily uses OpenRewrite recipes to automate updating your project's dependencies, source code, and documentation. These recipes cover many but not all of the items described in the link:https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0[Migration Guide 3.0].
```

not sure about this one but isn't the connotation of using "automatically"  that it will all be done for me; where "automate" is more a "make it easier for me" which seems like a better signal to send.

rolfedh replied:
Thanks. Updated. Good point.

rolfedh committed:
Create update guide

rolfedh commented:
@gsmet, I've updated the text based on review comments from you and Max. 
Would you confirm that this procedure works for Quarkus 2.7-based projects as well? I've updated the `2.13` in the text to `2.x`.

gsmet reviewed (approved):
Let's merge this, I will backport it today too.

Thanks.
